File: src/main/java/chat/teco/tecochat/auth/presentation/argumentresolver/AuthArgumentResolver.java
package chat.teco.tecochat.auth.presentation.argumentresolver;

import chat.teco.tecochat.auth.Auth;
import chat.teco.tecochat.auth.domain.Authenticator;
import lombok.RequiredArgsConstructor;
import org.springframework.core.MethodParameter;
import org.springframework.stereotype.Component;
import org.springframework.web.bind.support.WebDataBinderFactory;
import org.springframework.web.context.request.NativeWebRequest;
import org.springframework.web.method.support.HandlerMethodArgumentResolver;
import org.springframework.web.method.support.ModelAndViewContainer;

@RequiredArgsConstructor
@Component
public class AuthArgumentResolver implements HandlerMethodArgumentResolver {

    private static final String AUTH_HEADER_NAME = "name";

    private final Authenticator authenticator;

    @Override
    public boolean supportsParameter(MethodParameter parameter) {
        return parameter.hasParameterAnnotation(Auth.class);
    }

    @Override
    public Long resolveArgument(
            MethodParameter parameter,
            ModelAndViewContainer mavContainer,
            NativeWebRequest webRequest,
            WebDataBinderFactory binderFactory
    ) {
        String name = webRequest.getHeader(AUTH_HEADER_NAME);
        if (name == null) {
            throw new IllegalStateException("인증에 실패하였습니다.");
        }
        return authenticator.authenticateWithBase64(name).id();
    }
}

File: src/main/kotlin/chat/teco/tecochat/security/AuthArgumentResolver.kt
package chat.teco.tecochat.security

import chat.teco.tecochat.domain.Authenticator
import org.springframework.core.MethodParameter
import org.springframework.stereotype.Component
import org.springframework.web.bind.support.WebDataBinderFactory
import org.springframework.web.context.request.NativeWebRequest
import org.springframework.web.method.support.HandlerMethodArgumentResolver
import org.springframework.web.method.support.ModelAndViewContainer

private const val AUTH_HEADER_NAME = "name"

@Component
class AuthArgumentResolver(
    private val authenticator: Authenticator
) : HandlerMethodArgumentResolver {

    override fun supportsParameter(parameter: MethodParameter): Boolean {
        return parameter.hasParameterAnnotation(Auth::class.java)
    }

    override fun resolveArgument(
        parameter: MethodParameter,
        mavContainer: ModelAndViewContainer?,
        webRequest: NativeWebRequest,
        binderFactory: WebDataBinderFactory?
    ): Long {
        val name = webRequest.getHeader(AUTH_HEADER_NAME)
            ?: throw IllegalStateException("인증에 실패하였습니다.")
        return authenticator.authenticateWithBase64(name).id()
    }
}

